File: Sources/MonitorPresentation/MonitorZoomScale.swift
import Foundation

/// Presentation-only logarithmic lens ring. All limits and stops come from the
/// camera adapter; this type knows nothing about lenses or command encodings.
public struct MonitorZoomScale: Equatable, Sendable {
    public let minimum: Double
    public let maximum: Double
    public static let angularSpan = 210.0 * Double.pi / 180
    public static let tickIncrement = 0.01
    /// Equal-angle minor ticks across the full ring. Only labels use hundredths.
    public static let minorTickCount = 18
    public static let labeledTicks = [1.0, 1.5, 2.0, 3.0, 4.0, 6.0, 9.0, 12.0]
    /// Integer majors for visual marks and haptic crossings.
    public static let wholeStops = [1.0, 2.0, 3.0, 4.0, 6.0, 9.0, 12.0]

    public static func minorTickPositions() -> [Double] {
        (0...minorTickCount).map { Double($0) / Double(minorTickCount) }
    }

    public init(minimum: Double, maximum: Double) {
        self.minimum = minimum.isFinite && minimum > 0 ? minimum : 1
        self.maximum = maximum.isFinite ? max(self.minimum, maximum) : self.minimum
    }

    public func position(_ value: Double) -> Double {
        guard maximum > minimum, value.isFinite else { return 0 }
        return log(min(maximum, max(minimum, value)) / minimum) / log(maximum / minimum)
    }

    public func value(at position: Double) -> Double {
        guard position.isFinite else { return minimum }
        return minimum * pow(maximum / minimum, min(1, max(0, position)))
    }

    public func dragged(from value: Double, angleDelta: Double, current: Double? = nil)
        -> Double
    {
        _ = current
        guard angleDelta.isFinite else { return self.value(at: position(value)) }
        return self.value(at: position(value) - angleDelta / Self.angularSpan)
    }

    public func quantized(_ value: Double) -> Double {
        guard value.isFinite else { return minimum }
        let clamped = min(maximum, max(minimum, value))
        let snapped = (clamped / Self.tickIncrement).rounded() * Self.tickIncrement
        return min(maximum, max(minimum, snapped))
    }

    public func dialLabel(_ value: Double) -> String {
        String(format: "%.2f×", quantized(value))
    }

    public func isLabeledTick(_ value: Double, marks: [Double] = Self.labeledTicks) -> Bool {
        let tick = quantized(value)
        return marks.contains { abs(quantized($0) - tick) < Self.tickIncrement / 2 }
    }
}

/// One pointer lifetime owns one continuous-edit transaction. A changed viewport
/// or lens range cancels that lifetime; old pointer coordinates cannot restart it.
public struct MonitorZoomDrag: Equatable, Sendable {
    public private(set) var anchor: Double?
    private var cancelled = false

    public init() {}

    /// True only when a new edit transaction begins.
    public mutating func begin(at value: Double) -> Bool {
        guard !cancelled, anchor == nil, value.isFinite else { return false }
        anchor = value
        return true
    }

    /// End the current edit while rejecting updates until this pointer lifts.
    public mutating func cancel() -> Bool {
        let wasEditing = anchor != nil
        anchor = nil
        cancelled = true
        return wasEditing
    }

    /// UIKit may end or cancel a gesture. Both release its editing ownership.
    public mutating func end() -> Bool {
        let wasEditing = anchor != nil
        anchor = nil
        cancelled = false
        return wasEditing
    }
}

File: Apps/Android/monitor-ui/src/main/kotlin/com/opencapture/monitorui/MonitorZoomScale.kt
package com.opencapture.monitorui

import kotlin.math.abs
import kotlin.math.ln
import kotlin.math.pow
import kotlin.math.roundToInt

/** Continuous logarithmic input; hundredths are reserved for labels and accessibility. */
object MonitorZoomScale {
    const val ANGULAR_SPAN = 210.0 * Math.PI / 180.0
    const val TICK_INCREMENT = 0.01
    const val MINOR_TICK_COUNT = 18
    val labeledTicks = listOf(1.0, 1.5, 2.0, 3.0, 4.0, 6.0, 9.0, 12.0)
    val wholeStops = listOf(1.0, 2.0, 3.0, 4.0, 6.0, 9.0, 12.0)

    fun minorTickPositions(): List<Double> =
        (0..MINOR_TICK_COUNT).map { it.toDouble() / MINOR_TICK_COUNT }

    fun quantized(value: Double, minimum: Double = 1.0, maximum: Double): Double {
        if (!value.isFinite()) return minimum
        val lo = if (minimum.isFinite() && minimum > 0) minimum else 1.0
        val hi = if (maximum.isFinite()) maxOf(lo, maximum) else lo
        val clamped = value.coerceIn(lo, hi)
        val snapped = (clamped / TICK_INCREMENT).roundToInt() * TICK_INCREMENT
        return snapped.coerceIn(lo, hi)
    }

    fun dialLabel(value: Double, minimum: Double = 1.0, maximum: Double): String =
        String.format(java.util.Locale.US, "%.2f×", quantized(value, minimum, maximum))

    fun isLabeledTick(value: Double, minimum: Double = 1.0, maximum: Double): Boolean {
        val tick = quantized(value, minimum, maximum)
        return labeledTicks.any {
            abs(quantized(it, minimum, maximum) - tick) < TICK_INCREMENT / 2
        }
    }

    fun position(value: Double, minimum: Double, maximum: Double): Double {
        val lo = if (minimum.isFinite() && minimum > 0) minimum else 1.0
        val hi = if (maximum.isFinite()) maxOf(lo, maximum) else lo
        if (hi <= lo || !value.isFinite()) return 0.0
        return ln(value.coerceIn(lo, hi) / lo) / ln(hi / lo)
    }

    fun valueAt(position: Double, minimum: Double, maximum: Double): Double {
        val lo = if (minimum.isFinite() && minimum > 0) minimum else 1.0
        val hi = if (maximum.isFinite()) maxOf(lo, maximum) else lo
        if (!position.isFinite()) return lo
        return lo * (hi / lo).pow(position.coerceIn(0.0, 1.0))
    }

}
